### PR TITLE
[Clang importer + macros] Handle name lookup and type checking for expanded macros

### PR DESCRIPTION
**Pull request by DougGregor**
Introduce a number of fixes to allow us to fully use declarations that are produced by applying a peer macro to an imported declarations.

These changes include:
* Treating the source file containing `swift_attr` macros as one that stores macros, so we don't parse it as top-level code
* Ensuring that we have the right set of imports in the source file containing the macro expansion, because it depends only on the module it comes from
* Ensuring that name lookup looks in that file even when the DeclContext hierarchy doesn't contain the source file (because it's based on the Clang module structure)

Expand testing to be sure that we're getting the right calls, diagnostics, printing APIs, and generated IR symbols.

### CODE REVIEW

**Comment by DougGregor** [comment]
@swift-ci please smoke test

File: lib/IRGen/IRGenDebugInfo.cpp [hunk 1]
```diff
@@ -298,7 +298,9 @@ class IRGenDebugInfoImpl : public IRGenDebugInfo {
       if (auto generatedInfo = SM.getGeneratedSourceInfo(BufferID)) {
         // We only care about macros, so skip everything else.
         if (generatedInfo->kind != GeneratedSourceInfo::ReplacedFunctionBody &&
-            generatedInfo->kind != GeneratedSourceInfo::PrettyPrinted)
+            generatedInfo->kind != GeneratedSourceInfo::PrettyPrinted &&
+            generatedInfo->kind != GeneratedSourceInfo::DefaultArgument &&
+            generatedInfo->kind != GeneratedSourceInfo::Attribute)
```
Comment (hnrklssn) [inline lib/IRGen/IRGenDebugInfo.cpp:303]: The comment above says "We only care about macros, so skip everything else", but aren't macros attributes in the grammar? I wouldn't expect us to skip `GeneratedSourceInfo::Attribute` here.

File: test/Macros/print_clang_expand_on_imported.swift [hunk 2]
```diff
@@ -7,8 +7,10 @@
 // Build the macro library to give us access to AddAsync.
 // RUN: %target-swift-frontend -swift-version 5 -emit-module -o %t/macro_library.swiftmodule %S/Inputs/macro_library.swift -module-name macro_library -load-plugin-library %t/%target-library-name(MacroDefinition)
 
-// RUN: %target-swift-ide-test(mock-sdk: %clang-importer-sdk) -print-module -print-implicit-attrs -source-filename %s -module-to-print=CompletionHandlerGlobals -I %t -function-definitions=false -load-plugin-library %t/%target-library-name(MacroDefinition) -import-module macro_library -enable-experimental-feature MacrosOnImports > %t/imported.printed.txt
+// RUN: %target-swift-ide-test(mock-sdk: %clang-importer-sdk) -print-module -print-implicit-attrs -source-filename %s -module-to-print=CompletionHandlerGlobals -I %t -function-definitions=false -load-plugin-library %t/%target-library-name(MacroDefinition) -import-module macro_library -enable-experimental-feature MacrosOnImports > %t/imported.printed.txt 2> %t/stderr.log
 // RUN: %FileCheck -input-file %t/imported.printed.txt %s
+// RUN: echo "" >> %t/stderr.log
+// RUN: %FileCheck -check-prefix DIAGS -input-file %t/stderr.log %s
```
Comment (hnrklssn) [inline test/Macros/print_clang_expand_on_imported.swift:13]: Is it not possible to instead have `swift-ide-test` return an error code if the input has errors? That seems like a simpler solution. Not sure what effect they have, but it sounds like `--test-input-complete` or ` --typecheck` could be useful.

File: lib/IRGen/IRGenDebugInfo.cpp (hunk 1 above)
Comment (DougGregor) [reply lib/IRGen/IRGenDebugInfo.cpp:303]: The comment means "we only care about macro expansion buffers." I'll clarify it

File: lib/ASTGen/Sources/ASTGen/SourceFile.swift [hunk 3]
```diff
@@ -122,7 +122,8 @@ public func parseSourceFile(
       .preambleMacroExpansion,
       .replacedFunctionBody,
       .prettyPrinted,
-      .defaultArgument:
+      .defaultArgument,
+      .attribute:
```
Comment (rintaro) [inline lib/ASTGen/Sources/ASTGen/SourceFile.swift:126]: Could you move `.attribute` to `. memberAttributeMacroExpansion` branch below, it's still not implemented and doing the same thing, but just not to forget to move it later.

**Comment by rintaro** [comment]
While you are here, could you make the parsing to use `ParseSourceFileRequest`? (rdar://139119159)

File: lib/Parse/ParseRequests.cpp [hunk 4]
```diff
@@ -364,7 +366,8 @@ SourceFileParsingResult parseSourceFile(SourceFile &SF) {
       break;
     }
 
-    case GeneratedSourceInfo::MemberAttributeMacroExpansion: {
+    case GeneratedSourceInfo::MemberAttributeMacroExpansion:
+    case GeneratedSourceInfo::Attribute: {
       parser.parseExpandedAttributeList(items);
```
Comment (rintaro) [inline lib/Parse/ParseRequests.cpp:371]: Oh apparently, `swift_attr` accepts modifiers OR attributes, but `parseExpandedAttributeList` only accepts attributes? 🤔

Comment (DougGregor) [reply lib/Parse/ParseRequests.cpp:371]: Yeah, and we have some things that are permitted but are neither attributes nor modifiers, which is odd. I'm going to have to think about this refactoring more.

File: lib/ASTGen/Sources/ASTGen/SourceFile.swift (hunk 3 above)
Comment (DougGregor) [reply lib/ASTGen/Sources/ASTGen/SourceFile.swift]: Okay, done!

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Pull request merged**